Add support for Pico 2 (RP2350)

philipstarkey opened the pull request:
When built for the Pico 2, we now support:
- 60,000 instructions
- Clock speeds up to 150 MHz
- Running on PIO2

We've also:
- Added new `board` command that returns the board version (either `pico1` or `pico2`)
- Dynamically computed array sizes
- bumped the version to v1.2.0
- Refactored the cmake file to be a lot cleaner and support building for more than one board. The maximum number of instructions is now defined here
- Bumped the pico SDK version to v2.0.0
- Fixed a bug with Ubuntu 24.04 that broke the apt-mirror override in the docker container
- Updated the docker compose container for simultaneous build of firmware for multiple boards (in a way that doesn't cause long docker build times). This is also following the best practices outlined by the pico-sdk that encourages you to separate build dirs for different boards to avoid conflicts with cached cmake configuration.
- Removed the `pico_sdk_import.cmake` from the repo so it can be kept up-to-date with the version in the pico-sdk.

philipstarkey committed:
Remove `pico_sdk_import.cmake` from repo and instead copy it at build time from the SDK so it is kept up to date

philipstarkey committed:
Update PrawnBlaster firmware to support the new Pico 2 (RP2350) board.

When built for the Pico 2, we now support:
- 60,000 instructions
- Clock speeds up to 150 MHz
- Running on PIO2

We've also:
- Added new `board` command that returns the board version (either `pico1` or `pico2`)
- Dynamically computed array sizes
- bumped the version to v1.2.0
- Refactored the cmake file to be a lot cleaner and support building for more than one board. The maximum number of instructions is now defined here
- Bumped the pico SDK version to v2.0.0
- Fixed a bug with Ubuntu 24.04 that broke the apt-mirror override in the docker container
- Updated the docker compose container for simultaneous build of firmware for multiple boards (in a way that doesn't cause long docker build times). This is also following the best practices outlined by the pico-sdk that encourages you to separate build dirs for different boards to avoid conflicts with cached cmake configuration.

philipstarkey committed:
Added new `board` command to documentation

philipstarkey committed:
Correct comment

dihm commented on docker/Dockerfile:
```diff
@@ -6,21 +6,23 @@ ARG APT_MIRROR="mirror://mirrors.ubuntu.com/mirrors.txt"
+    apt install -y cmake gcc-arm-none-eabi libnewlib-arm-none-eabi libstdc++-arm-none-eabi-newlib build-essential
 
 # Install Pico SDK into a new stage
 FROM base as buildtools
```
docker complains that the command words to not match case (ie it wants `FROM base AS buildtools`)

dihm commented on docker/Dockerfile:
```diff
@@ -6,21 +6,23 @@ ARG APT_MIRROR="mirror://mirrors.ubuntu.com/mirrors.txt"
     mkdir -p /pico/ && \
     cd /pico/ && \
-    git clone https://github.com/raspberrypi/pico-sdk.git --branch 1.5.1 && \
+    git clone https://github.com/raspberrypi/pico-sdk.git --branch 2.0.0 && \
```
I went ahead and had a pass at bumping this to 2.1.1 (current release). I got errors about not installing the correct version of picotool (which is normally done automatically). It would be nice to sort that out eventually, but we can probably pass for now.

Note: I suspect the fix might be explicitly installing picotool from source, instead of relying on the SDK to do it for us.

dihm reviewed (approved):
Finally got some initial testing done here. I can confirm that both a Pico1 and 2 work with this build of the firmware in a basic test.

While it would be nice to apply a more thorough test, I don't really have time for a full battery to confirm specs ATM. I am confident enough to merge and release (after minor comments are at least looked at) and deal with more thorough tests later.

dihm committed:
Make capitalization consistent in Dockerfile command

dihm committed:
Update pico-sdk to 2.1.1

Also manually install picotool from source to ensure its version
matches the sdk version (and doesn't have to be rebuilt by each
firmware build). This was necessary to bump pico-sdk to 2.1.1

dihm commented:
I've gone ahead and fixed my two comments and bumped the SDK to 2.1.1. Both Pico1 and 2 are nominally functional built under this version of the SDK. More rigorous testing would be nice, but I'm happy to merge as-is and save the testing for later when we allow for the RP2040 to get to 200MHz (which should be a separate PR).

philipstarkey committed:
Add warning regarding the RP2350 E9 errata
